rfcs/0000-a-forth-userspace.md

```diff
@@ -0,0 +1,72 @@
+# RFC - Forth Userspace
+
+## Goal
+
+In order to make early prototyping and iterating on the design of mnemos **easier**, this RFC proposes the
+addition of a single kind of "userspace" environment, a `forth` vm that can act as a program
+or interactive shell interface. This RFC also proposes a minimal set of capabilities and interfaces
+necessary to make this possible.
+
+## Background
+
+mnemos has some design around what it means to launch kernel services, but has no kind of userspace
+separate from the kernel itself. Designing this "properly" requires a lot of subtle decisions to be
+made, so this rfc proposes to ignore these problems until we are much smarter people later in the
+project.
+
+James has been working on an extensible forth vm written in Rust, and recently Eliza has been helping
+to expand the vm, including the ability to have async host-provided intrinsic functions. While limited,
+this environment is likely suitable for two main things:
+
+* "gluing together syscalls" to test out kernel capabilities and bring up hardware support
+* Crossing the "embedded system"/"computer" rubicon, allowing for runtime developed programs
+
+## The proposal
+
+* The OS will spawn a single userspace entity at boot time
+    * This entity will be a cooperative async task running an interactive forth vm
+    * It will be capable of interacting with the userspace/kernel interface (sending/receiving kernel messages)
+* This "task 0" forth environment will be capable of spawning other userspace tasks
+    * It can only spawn other forth vms, running their own separate environments
+    * When spawning a new task, a snapshot of the current task's dictionary will be made and shared with the child environment
+    * All forth vm tasks will be cooperatively scheduled
+
+## Details
+
+### Changes to `forth3` - the current forth vm
```

Comment:
maybe worth linking my in progress branch on forth3 for this?

Reply:
Linked!